src/zerocoin_v3.cpp

```diff
@@ -604,51 +596,64 @@ void CSigmaTxInfo::Complete() {
 CSigmaState::CSigmaState() {
 }
 
-int CSigmaState::AddMint(
+bool CSigmaState::AddMintsToStateAndBlockIndex(
         CBlockIndex *index,
-        const sigma::PublicCoin &pubCoin) {
-    sigma::CoinDenomination denomination = pubCoin.getDenomination();
-
-    if (latestCoinIds[denomination] < 1)
-        latestCoinIds[denomination] = 1;
-    int	mintCoinGroupId = latestCoinIds[denomination];
-
-    // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
-    // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
-    // each block will end up in the same accumulator.
-    SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
-    int coinsPerId = ZC_SPEND_V3_COINSPERID;
-    if (coinGroup.nCoins < coinsPerId // there's still space in the accumulator
-        || coinGroup.lastBlock == index // or we have already placed some coins from current block.
-        ) {
-        if (coinGroup.nCoins++ == 0) {
-            // first group of coins for given denomination
-            assert(coinGroup.firstBlock == nullptr);
-            assert(coinGroup.lastBlock == nullptr);
-
-            coinGroup.firstBlock = coinGroup.lastBlock = index;
-        } else {
-            assert(coinGroup.firstBlock != nullptr);
-            assert(coinGroup.lastBlock != nullptr);
-            if (coinGroup.lastBlock->nHeight > index->nHeight) { 
-                assert(coinGroup.lastBlock->nHeight <= index->nHeight);
-            }
+        const CBlock* pblock) {
+
+    std::unordered_map<sigma::CoinDenomination, size_t> blockDenomCounts;
+    for (const auto& mint : pblock->sigmaTxInfo->mints) {
+        blockDenomCounts[mint.getDenomination()]++;
+    }
+
+    for (const auto& pubCoin : pblock->sigmaTxInfo->mints) {
+        sigma::CoinDenomination denomination = pubCoin.getDenomination();
+
+        if (latestCoinIds[denomination] < 1)
+            latestCoinIds[denomination] = 1;
+        int	mintCoinGroupId = latestCoinIds[denomination];
+
+        // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
+        // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
+        // each block will end up in the same accumulator.
+        SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
+        int coinsPerId = ZC_SPEND_V3_COINSPERID;
+        int coinsPerIdLimit = ZC_SPEND_V3_COINSPERID_LIMIT;
+        if ((coinGroup.nCoins < coinsPerId // there's still space in the accumulator
+            && coinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit)
```

Comment:
I think better to do this check for each denomination, not for each coin.

Reply:
good idea. thanks.